lib/rpmsg/rpmsg_virtio.c

```diff
@@ -610,6 +611,17 @@ int rpmsg_init_vdev(struct rpmsg_virtio_device *rvdev,
 		    rpmsg_ns_bind_cb ns_bind_cb,
 		    struct metal_io_region *shm_io,
 		    struct rpmsg_virtio_shm_pool *shpool)
+{
+	return rpmsg_init_vdev_with_config(rvdev, vdev, ns_bind_cb, shm_io,
+					   shpool, NULL);
+}
+
+int rpmsg_init_vdev_with_config(struct rpmsg_virtio_device *rvdev,
+				 struct virtio_device *vdev,
+				 rpmsg_ns_bind_cb ns_bind_cb,
+				 struct metal_io_region *shm_io,
+				 struct rpmsg_virtio_shm_pool *shpool,
+				 const struct rpmsg_virtio_config *config)
```

Comment:
can we pass the rx/tx size directly to simplify the caller to define a struct rpmsg_virtio_config? and then change the function name to rpmsg_init_vdev_with_size.

Reply:
The definition I proposed is easily extensible if we find more configuration options to add in the future. We would just need to add more fields to the structure, instead of creating a new function for each potential configuration options permutations.

But if specialized init function for overriding sizes only is preferable, I can modify this implementation.

Reply:
Agree with @hubertmis, it would be better to keep a structure for future extension..
I just wonder about naming ( but no strong opinion) could we use "config" or "attributes"

Reply:
`const struct rpmsg_virtio_config *config `:
If the config is retrieved from the resource table could not be const. For instance regarding https://github.com/OpenAMP/open-amp/pull/155/

Reply:
> > If the config is retrieved from the resource table could not be const
> 
> `rpmsg_init_vdev_with_config()` does not modify the content of the structure. C allows assigning non-const pointers to const pointer types like:
> 
> ```
> struct rpmsg_virtio_config modifiable_config;
> modifiable_config.txbuf_size = condition ? 100 : 256;
> modifiable_config.rxbuf_size = 512;
> rpmsg_init_vdev_with_config(..., &modifiable_config);
> ```

Could you confirm this by a test?  Because a simple test on my side, adding const to a function parameter, generates a " discards ‘const’ qualifier" warning on build.


> 
> If `VIRTIO_RPMSG_F_BUFSZ` was implemented like in #155, I think it would modify the buffer `config` stored inside `struct rpmsg_virtio_device`, not the one passed as an argument to the initialization function (which is copied to `struct rpmsg_virtio_device` during initialization).
> 
> I don't see the reason to remove `const` from the new parameter type. Am I missing something?

Only Build warning and MISRA concerns.

Reply:
I've tested it with a sample Zephyr application:
```
        struct rpmsg_virtio_config config = {
            .txbuf_size = 60,
            .rxbuf_size = 160,
        };
        err = rpmsg_init_vdev_with_config(&ctx->rvdev, &ctx->vdev, ns_bind_cb,
                      ctx->shm_io, &ctx->shpool, &config);
```
C99 6.3.2.3:
> For any qualifier q, a pointer to a non-q-qualified type may be converted to a pointer to the q-qualified version of the type; the values stored in the original and converted pointers shall compare equal.

MISRA C 8.13

What code are you using that generates a warning? I'm confused, because C standard allows this conversion.

Reply:
My apologies, my quick and dirty test adding a const was not good.
By applying your patch and updating system/linux/machine/generic/platform_info.c for testing: no warning.